src/components/Histoslider.js

```diff
@@ -2,48 +2,72 @@ import React, { Component } from "react";
 import PropTypes from "prop-types";
 import { max, min } from "d3-array";
 import { scaleLinear as linear } from "d3-scale";
-
+import { debounce } from 'lodash';
 import Histogram from "./Histogram";
 import Slider from "./Slider";
 
 const SLIDER_HEIGHT = 30;
 
 class Histoslider extends Component {
-  constructor() {
-    super();
+  constructor(props) {
+    super(props);
     this.state = {
-      dragging: false
+      dragging: false,
+      selection: this.props.selection
     };
+
+    this.debouncedChange = debounce(this.props.onChange, 300)
   }
 
   dragChange = dragging => {
     // TODO - debounce
     this.setState({ dragging });
   };
 
+  componentDidUpdate(prevProps){
+    if (prevProps.onChange !== this.props.onChange){
+      this.debouncedChange = debounce(this.props.onChange, 300)
+    }
+    if (this.props.selection && this.props.selection.length === 0) {
+        const sortedData = this.props.data.sort((a, b) => +a.x0 - +b.x0);
+        const extent = [
+          min(sortedData, ({ x0 }) => +x0),
+          max(sortedData, ({ x }) => +x)
+        ];
+        this.setState({selection: extent});
+        this.props.onChange(extent)
+      }
+  }
+
```

Comment:
This should add a `componentWillUnmount` so it can cancel any pending debounced calls:

```js
componentWillUnmount() {
  this.debouncedChange.cancel();
}
```